### DISCUSSION TITLE
3D Printed Enclosure; Optimization/Redesign Parameters

### DISCUSSION THREAD

**Discussion by GreenOrpheus** [geezacoleman/OpenWeedLocator#19] [General]
Hello!

I'd like to take on the task of redesigning the 3D printed enclosure, but need a bit of guidance on the design philosophy you shot for in the original, and how much of that philosophy you'd like to keep going going forward.

**1. Is an enclosure even needed?** 
I have all the pieces of of the OWL inside a weatherproof project box with a transparent lid, [similar to this one.](https://www.ubuy.com.pl/en/product/11806NY0-gratury-junction-box-hinged-cover-transparent-lid-ip65-waterproof-plastic-enclosure-for-electrical-p) The components are mounted on a 3d printed platform to hold them in place, and holes are drilled through the project box for the cable connections, lights, ventilation nubs, et. These boxes are not cheap, (mine was ~$60 CAD), but there are many different types, and I'm sure I could find one for half the price if I searched hard enough. It also solves the problem of my next question: environmental infiltration.

**2. How will the components be protected from the environment?**
As I haven't yet put my prototype into action, I don't know how well it will react to environmental factors like dust, humidity, spray drift, heat, et. As I understand it, the current design just has an open hole for the camera lens to poke through. How did it stand up to field tests this last year? Are there any ideas on how to mitigate environmental damage? I like to put my electronic gizmos inside project boxes, but I am unsure how the transparent lid might affect camera performance in the field.
Another alternative would be to design the case to accept some sort of [UV filter](https://www.amazon.ca/Polaroid-Optics-52mm-Protective-Filter/dp/B00HFUU6RU/ref=sr_1_6?crid=2ULAESRK6ZH9W&keywords=rugged+uv+filter&qid=1643319109&sprefix=rugged+uv+filt%2Caps%2C318&sr=8-6) like is commonly put on the ends of camera lenses for cheap protection.

**3. How modular/customizable should the enclosure be?**
The current enclosure has plenty of room for swapping out slightly different parts. I can make the enclosure compact, but I may sacrifice some customizability in the process. Is this ok, or should I shoot for more modularity? A parametric model would be ideal, but with my current software I am unable to create one.

**4. Is the current parts list the one you'd like me to stick to?**
My own build differed slightly from the parts list. For example I was unable to get the preferred relay board in Canada, and so went with a cheap and widely available alternative. This alternative had slightly different dimensions/hole locations. As another example, the Bulgin connectors were quite expensive, so I went with a cheaper 6 pin connector I had used in the past. Has any thought been given to refining/changing the parts list, or should I stick to what is currently listed?

**5. How important are the camera and bolt hole locations?**
Is it necessary to keep the camera the current distance away from the rear of the enclosure?
Is it necessary to keep the bolt hole locations in their current spot? 

I'm sure I'll have more questions as I iterate further. I hope to post of picture of my setup soon, when the parts arrive to repair my cellphone camera.

**Comment by geezacoleman** [comment]
This is great! Thanks for taking it on, very excited to see what you come up with. These are some great questions to start, so I've had a go answering them the best I can. @williamtsalter may also have some comments here. Before I answer in more detail, certainly feel free to just run with any design or idea you have; don't feel limited by my responses if there's a better option. The main philosophy here is open collaboration, ease of use/assembly and modularity. Beyond those very loose but important ideas, just do what you think is best.

_On to the specifics..._
A bit of context on the current enclosure. 3D design is definitely not my strength, so the current case is a very basic prototype I've been meaning to update that was designed with the following in mind:
* printable on my Prusa MK3S
* easily swappable components (RPi, relay board, camera) for upgrades/rapid changes during testing
* adjustable camera mounts for different camera types (HQ cam vs v2 vs NOIR)
* general protection from rain, dust, drops, bumps
* ability to mount to frames with M6 bolts
* simple to assemble

**1. Is an enclosure even needed?**
Those plastic enclosures are great for a few and I've made previous versions in those too, but needing to drill holes multiple times increases the time/tools needed and makes doing lots of them quite laborious. Whereas if you can send a few .stl files for printing on your own machine or via an online service, it makes it easier to assemble and be confident it will fit once they are done.

**2. How will the components be protected from the environment?**
This is fairly critical and the camera opening as it stands is not conducive keeping things out. We've had two mounted on a robot since April 2021 and they have been used quite a few times with no issues in Australian heat/some rain after we put tape over the rear join to seal it. But once it's on faster moving equipment then it will obviously start causing problems.

What I have done in the past for field cameras is use small bits of glass that are glued in place. E.g. use a [circular bit of glass](https://www.etsy.com/dk-en/listing/809083162/82040pcs-flat-clear-glass-tile-round?click_key=5ec22bdeacf84f831393f36b54ff5d2fecb815b0%3A809083162&click_sum=4b694f15&ga_order=most_relevant&ga_search_type=all&ga_view_type=gallery&ga_search_query=30mm+glass+tile&ref=sr_gallery-1-9), with a recessed area in the design that neatly fits the glass. It would be fairly easy to add this into the assembly guide too and the glass is quite cheap. One idea I've been considering is active dust removal via a fan directing a sheet of air over the glass. Not sure this would actually work in practice, but 3D printing makes creating inbuilt ducting a bit easier.

The other consideration is room for a rubber bead/seal that can used to prevent moisture getting in. This also leads to cooling issues, but if we mount a heat sink inside that is thermally connected to the outside/mounting frame this could help move heat out.

**3. How modular/customizable should the enclosure be?**
Modularity is fairly key and enclosure should ideally have room for upgrades to say Jetson Nano, Xavier NX etc for green-on-green work, which is in the pipeline. The current design is wider than the Pi because of this. On using stands/mounting plates - stands means you can simply reprint mounting plates if you want to upgrade. In saying that, current amount of space is quite excessive, and there would be more efficient ways to maintain similar levels of modularity without wasting space. See what you come up with here! 

**4. Is the current parts list the one you'd like me to stick to?**
Parts list beyond the RPi and camera is quite fluid. It's mainly a guide, I've built them with different Pololu power converters/relay boards as well and just changed hole locations by drilling or redesigning the plates. There seem to be so many variations out there that it's difficult to design for all, hence the importance of modularity and having some room to move too. If you have some better plugs definitely put those up, but also include a version with just the cable glands as we did. They are more universal, cheaper and easier/faster to put together than soldering terminals.

**5. How important are the camera and bolt hole locations?**
Distance between camera and nozzle is not _currently_ included in any of the software but addition of GPS would make this important. I would stick to a whole number distance for ease of measurement but beyond that, no worries.

I would stick to M6 size holes, but mounting distance/dimensions isn't set on anything specific.

If you were to design a unit for a single set of components and known end use case for your own purposes, the end design could be much more efficient/compact. But by making it accessible to people everywhere with modularity, it helps more people use it. So you could have your own optimised design for your units and then a separate modular one, all very much up to you though!

Even a simple update of the current version is fantastic. Keep the questions coming as well!

**Reply by GreenOrpheus** [reply]
Thanks! That gave me a good place to start on designing. Here's where I'm at after a few iterations. I may have overengineered things just a bit...

I made a box with vertical dovetail rails on the interior sides. There are four rail slots, for up to four potential 50mm x 100mm "bays." Bays can be combined to hold larger items depending on the size of "holder" The holders are simple plates that can be configured with screw hole locations by the user to secure their specific parts. the box has room for two layers of holders, stacked on top of each other. Each holder layer can fit parts up to 33mm tall. Wiring allowances are made in between the rails, and in a small layer underneath the bays. I don't expect these wiring allowances will need to be used, as there should be ample room to run wires alongside the parts, or through the holders. Each holder is secured to the rails with 4 M2.5x16 screws threaded directly into the plastic. Since we have to work with M2.5 for the raspberry pi anyway, I figured there's a good chance people will have these laying around. I've been using that size to secure all my parts together.

This design easily accommodates all the parts of my setup, and even fits the Jetson Nano Dev Kit. The relays I've been using are 50mm wide, so they fit perfectly in a single bay. I think the recommended ones on your parts list are 55mm wide, so they'll need to be placed on a double wide holder. Not really a problem, as you can cram other parts onto that holder as well.

There are still lots of things missing from this design. I haven't settled on the M6 bolt hole locations yet, but I'm leaning towards moving them onto brackets on the exterior of the box. I have not designed a lid or a holder for the camera bay, but those should be done soon. Also each holder is just a blank frame with no part specific screw holes yet.

There are still some tests to be done. My printer has been acting up, so haven't been able to print a full box, only 2 test frame bays. Everything has worked well so far, but I won't know for sure until a full assembly is complete. Also, I haven't done any wiring on these tests. Once it all comes together there may be assembly issues I'm not anticipating.

Looking for feedback on the basic concept. Do you think it's too complex, too finicky? Any questions? Let me know


![DSCF3597-1](https://user-images.githubusercontent.com/98173678/152441326-3c77e0d5-43cd-4b8e-b8a0-c3685f7f32e5.jpg)
![DSCF3598-1](https://user-images.githubusercontent.com/98173678/152441334-55f6458f-fcc3-40e4-8cbc-28caad1b2972.jpg)
![DSCF3599-2](https://user-images.githubusercontent.com/98173678/152441341-231308da-4531-46da-982d-fa56dfde69da.jpg)
![DSCF3600-3](https://user-images.githubusercontent.com/98173678/152441350-a258b023-5711-45b0-a933-5b22d4b8c39e.jpg)
![DSCF3601-4](https://user-images.githubusercontent.com/98173678/152441367-dcb5855d-15f8-473b-a456-a4472e3f399e.jpg)
![DSCF3602-5](https://user-images.githubusercontent.com/98173678/152441383-03842e00-27bc-4284-b581-2d4818c287eb.jpg)
![DSCF3603-6](https://user-images.githubusercontent.com/98173678/152441392-26288770-6932-430e-a1db-4bd6fae0bc9f.jpg)
![DSCF3604-7](https://user-images.githubusercontent.com/98173678/152441403-79beed3a-b5ed-45a6-9227-84dac84ec29b.jpg)
![V04_ISO Layers](https://user-images.githubusercontent.com/98173678/152441419-13a91c34-3cb3-401d-98a0-621b3169ca5b.jpg)
![V04_Plan](https://user-images.githubusercontent.com/98173678/152441424-da2f13ab-7f7f-406e-a634-a9510c17c6d9.jpg)
.

**Reply by geezacoleman** [reply]
That's a really neat, modular design you've come up with! I like the way you can add more 50 mm wide blocks together for larger pieces. It would make it quite easy to add a power bay, accessories bay, fan or other pieces. Nice work!

A few questions below:
* how do you see the camera being mounted? Would that just be a specialised mounting bay that could be redesigned for each camera?
* Will you add in mounting holes on the base pieces for each component? Ideally you'd want to be able to install this without drilling anything.
* Could you add a slot for a hex/square nut in the uprights so you can tighten down each bay more securely? Just in the plastic might make them fall out with lots of vibration/use over time.
* How long does it take to print a single bay piece?
* What are the overall dimensions of the new setup?

The other concern is around the strength of those uprights and pieces in general. The original camera mounts were very fragile, and while these dovetails would be supported by the dovetailing and locking screws when mounted in the box, I'd be worried they might break during construction. Is there a way you could brace them a little and add some cross bracing on the bases as well? They might also be strong enough as is, hard to say exactly from images alone.

Considering the mounting dovetails will be part of the base piece, I don't see that being too much of an issue for excessive complexity or anything, it looks fairly straightforward! Locking pins are in a much more accessible location too.

**Reply by geezacoleman** [reply]
There were a few others that were discussing the enclosure design in #8 @merbeinrobotics @plains203 @rfirth-code @DaleKirby that might be interested in this discussion too. 

Any thoughts on experience with old design and how this new design might help or could be changed?

**Reply by GreenOrpheus** [reply]
Hey, I'm back. Still chipping away at this. Here's V06.

The strength of the uprights on the holders seems good initially. I can give these pieces a lot of abuse before they break. I specifically beefed up the camera one as well. We'll see how it holds up in testing. I've changed the lid/base arrangement to be similar to your original design, as I was afraid assembly would be difficult reaching down into an enclosed box. This reduces the strength of the enclosure upright rails, but I've also added captive nuts to fix them rigidly to the lid, top and bottom.

At this point I think I've added everything on my checklist except for a transparent viewport integration for the camera. Still running with the ol' hole.

- **How do you see the camera being mounted? Would that just be a specialised mounting bay that could be redesigned for each camera?** Yes. I've designed one for the HQ camera (pictured), and could whip up more for future camera recommendations.

 
- **Will you add in mounting holes on the base pieces for each component? Ideally you'd want to be able to install this without drilling anything.** I'll make up holders with mounting hole for the recommended equipment list. The ones in the top left of the picture are for my equipment, just so you can get an idea.

- **Could you add a slot for a hex/square nut in the uprights so you can tighten down each bay more securely? Just in the plastic might make them fall out with lots of vibration/use over time.** I've now added captive nuts for the holders, and for the lid.

- **How long does it take to print a single bay piece?** Print times with a Prusa MK3S - .4 nozzle - .30mm layer - 20% infill:

     Enclosure Bottom - ~8.5h
     Enclosure Lid - ~6.25h

     100x100 Holder (blank) - ~1.75h
     50x100 Holder (blank - ~1.25h

     100x100 Holder (with holes) ~1.5h
     50x100 Holder (with holes) - ~1h
     50x100 HQ Camera Holder - ~1.5h

- **What are the overall dimensions of the new setup?** I didn't manage to shrink this as much as I'd like (and I think I even raised the height. Whoops.), but given the design I went with I think it's as small as I can make it at this point.

     Width (exterior mounting holes) - 156mm
     Width (interior mounting holes) - 120mm
     Length - 210mm
     Height - 80mm

There are still one or two tweaks to be made to this design, then it is ready to be tested. I have still not been able to test print the lid, or integrated bottom due to printer issues. My printer just doesn't like large objects. So my list at the moment is:

-Misc. tweaks
-Camera viewport thingy somehow
-Test print
-Test assembly
-Field testing 

I'm looking for another round of feedback/questions/suggestions if anyone has any.

![V06 Holders](https://user-images.githubusercontent.com/98173678/156896209-47fdfebc-4260-4da1-815c-563a788d43ac.jpg)
![20220305_124847](https://user-images.githubusercontent.com/98173678/156896428-0af90c61-892c-4d98-b369-96973ccdaa43.jpg)

**Reply by geezacoleman** [reply]
Nice work! Looking really good. It's great to see it coming along.

**Camera mount**
That camera mount looks quite nice! It might need to be rotated 90 degrees though so we have the right image dimensions. Could probably get away without connecting it with the tripod mount if I'm seeing that correctly in the photo.

**Vertical posts**
The long uprights at the front seem a little tall/exposed to damage when building - I know I dropped/bumped the original design I printed which broke the camera supports. We ended up redesigning away from tall vertical pillars in #11. What are the dimensions of those pillars?

We could also try printing a version too if you wanted to link some of the .stls and give it a test for you.

**Reply by GreenOrpheus** [reply]
The dimensions of the pillars for the corner of each bay are based on 6mm x 6mm squares, not counting the dovetail protrusions. So an even larger surface area in total. In my small test prints it was quite sturdy (I threw it down some stairs), but I'd need to print off the current version and do loaded tests to really know for sure.

My printer is now completely out of commission, with an escalating series of new exciting errors each time I go to fix the old ones. I've ordered some new parts, but I don't have a lot of hope. If someone could test print this design that would be amazing. Fingers crossed that I've got everything aligned properly. (https://we.tl/t-OCGZ1RHOp9) Link should be active for a week.

**Changes:** I've reduced the number of captive nuts & screws holding the lid to the base. Instead I've opted for alignment pins on the top to increase rigidity of the assembly. Some other minor tweaks.

![V07](https://user-images.githubusercontent.com/98173678/159297184-55164f9d-c1f1-49b1-b6aa-2460df26fff4.jpg)

**Reply by geezacoleman** [reply]
That seems a lot sturdier than our initial attempts at the camera mounts, but they were only a few mm square. Looking good!

We might have some difficulty doing it this week, but @williamtsalter may have more time next week or I should be back with the 3D printer the week after. So within the next few weeks we'll print out the design, take some photos and let you know!

**Reply by GreenOrpheus** [reply]
Thank you. If I make any further changes I'll post the updated STLs here.

**Reply by GreenOrpheus** [reply]
Repairs to the printer went better than expected. I replaced the PINDA sensor and everything works again! I managed to do a test of the bottom half of the case, and have discovered a couple weak points. I'll work on revisions. In the meantime I don't recommend printing this version.

**Comment by MerbeinRobotics** [comment]
Hi guys.  I've been noticing this discussion and it't great to see the developments with the enclosure.  I modified the original enclosure design by making it a bit shorter to fit on my Creality Ender 3 printer.  I hope to be doing more work on it soon but we are in the middle of harvest so I've been a bit busy.  With the price of herbicides this project is very relevant.  eg 20L of Glyphosate is about $280 at the moment!

**Reply by geezacoleman** [reply]
Thanks for the update @MerbeinRobotics! Yes the price of agrochemicals really makes spot spraying critical. Once you get it up and running it would be great to see it working!

We've just set up an [online workshop ](https://www.eventbrite.com/e/openweedlocator-live-assembly-and-demonstration-tickets-305069871197)on assembling/demonstrating the OWL so feel free to join as well.

**Comment by geezacoleman** [comment]
Another approach to the enclosure from a farmer in Ontario, Canada. Uses a DIN rail and aluminium enclosure for added ruggedness and ease of accessibility into the case.

From the [AgOpenGPS thread](https://discourse.agopengps.com/t/opensource-weed-recognition-and-image-database-weedai/6359/5) on the OWL/Weed-AI.

![image](https://user-images.githubusercontent.com/51358498/161270224-e8816bf4-7e7d-4c07-b7a3-70432bc7ff3c.jpg)

**Comment by geezacoleman** [comment]
So after some discussions summarised in #48 , spent some time designing the OWL. Focused on:
* compactness
* durability (i.e. can it fit inside a bit of steel and is it watertight)
* ease of assembly/attachment to the implement

End result:
Front | Back | Assembled | Lid on
:-------------------------:|:-------------------------:|:-------------------------:|:-------------------------:
 ![20220824_152855](https://user-images.githubusercontent.com/51358498/186535041-43542fbe-6a46-4b59-8918-5e56bc08b850.jpg)|![20220824_152906](https://user-images.githubusercontent.com/51358498/186535173-fd05c7c4-4881-4e5d-b5fc-650bb17e41b2.jpg)|![20220824_152740](https://user-images.githubusercontent.com/51358498/186535209-25bbb2ba-7dec-4f2b-8f12-634fd8d50c23.jpg)|![20220824_153349](https://user-images.githubusercontent.com/51358498/186535729-e7a36967-93be-46c0-811e-6689baa17cb2.jpg)

Only three printed pieces required for the design too, for a total of about 17-18 hours of printing (using 20% infill and 0.2mm quality setting on Prusa Slicer), compared to > 28 hours previously.

Draft [TinkerCAD design](https://www.tinkercad.com/things/2tCehwKUddq) is available for 3D printing if you want to get started, but these are very likely to change.

Major changes on previous design:
* removed lights and external fuse
* used PiRelay HAT instead of separate board to reduce wiring
* powered off pins 2 and 6 of the GPIO header board instead of USB-C
